[feat] 메이트 리스트 뷰와 산책 신청 뷰 연결 작업

kelly-chui opened the pull request:
### 🔖  Issue Number

close #79

### 📙 작업 내역

> 구현 내용 및 작업 했던 내역을 적어주세요.
> 
- 메이트 리스트 뷰, 산책 신청 뷰 연결
- 산책 신청 뷰 바인딩
- Supabase Database fetch 동작 개선 + 서버 동작 유즈케이스 추가
    - 메이트를 바인딩 할 때 필요해서 추가했으나, 메이트를 직접 파라미터로 받는 방법으로 변경되어서 현재 쓰이고 있진 않습니다.
    - 유즈케이스도 마찬가지로 더 서버에서 메이트 정보를 불러오지 않으므로 구현은 되어있으나, 더 이상 필요하지 않습니다. 하지만 `RequestMateInfoUseCase`는 `RequestUserInfoUseCase`와 통합할 수 있다고 예상합니다.
- 키워드 뷰 로직 개선 및 수정

https://github.com/user-attachments/assets/5beec319-290e-483c-8512-a8baa88b15b6

### 📝 PR 특이사항 

> PR을 볼 때 팀원에게 알려야 할 특이사항을 알려주세요.
> 
- 메이트 리스트 뷰와 산책 신청 뷰를 연결했습니다.
    - 메이트의 정보는 서버에서 받아오는 것이 아닌 메이트 리스트 뷰에서 가지고 있는 `Mate`의 정보를 그대로 파라미터로 받는 방식으로 구현했습니다.
```swift
static func createRequestWalkModule(mate: Mate) -> UIViewController {
    ...
    let interactor: RequestWalkInteractable = RequestWalkInteractor(
        mate: mate,
        requestWalkUseCase: requestWalkUseCase,
        requestProfileImageUseCase: requestProfileImageUseCase
    )
    let router: RequestWalkRoutable & RequestWalkBuildable = RequestWalkRouter()
```
- 산책 신청 뷰에서 키워드를 띄우는 로직을 개선했습니다.
    - `KeywordView`가 이니셜라이즈 될 때 레이블 텍스트 앞에 "#"을 붙여서, 바인딩 때 "#"이 출력되지 않는 문제가 있었습니다. 이를 `text`를 override 해서 해결했습니다.
    - 또한 `KeywordView`가 처음부터 빈 String으로 이니셜라이즈 된 상태여서 키워드가 1개인 경우에도 두 개가 보이는 문제가 있었습니다. 이 부분은 `UIView`의 `isHidden` 프로퍼티를 사용해서 해결했습니다.
    - 마지막으로 인덱스로 직접 접근하던 기존 방식에서 `zip`과 `forEach`를 사용해서 키워드 뷰에 레이블 텍스트를 넣을 수 있도록 개선했습니다. 
```swift
zip([firstKeywordView, secondKeywordView], keywords)
    .forEach { keywordView, keyword in
        keywordView.text = keyword
        keywordView.isHidden = false
    }
```
- Supabase
- 현재 팀원들이 진행하고 있는 부분에서 서로 UseCase 구현이 많이 겹칠 것으로 예상됩니다. 추후 몹 프로그래밍 할 때 통합해야 할 것 같습니다.

Commit:
[feat] #79 메이트 리스트 뷰와 산책 신청 뷰 연결

메이트 리스트 뷰에 있는 악세서리 셀 버튼을 이용해서 신청 뷰로 연결될 수 있도록 했습니다.

Commit:
[fix] #79 프로필 뷰 내부 키워드 라벨 뷰 수정

- 이니셜라이저 부분에 #을 추가하는 로직이 있어서 이 부분을 text를 오버라이드 받아서 해결하는 것으로 수정했습니다.
- 키워드 뷰에 직접 인덱스에 접근해서 레이블 텍스트를 추가하는 부분을 zip과 forEach 메소드를 사용해서 개선했습니다.

Commit:
[feat] #79 산책 신청 뷰 바인딩

- 기존에 번호로 메이트를 구분하는 부분에서 직접 메이트 객체를 받을 수 있도록 변경했습니다.
- 프로필 이미지를 서버에서 불러올 수 있도록 했습니다. 아직 유즈 케이스가 구현되지 않았습니다.

Commit:
[ui] #79 산책 신청 뷰 이펙트 수정

주변 화면의 밝기를 dark로 변경하여 산책 신청 뷰와 메이트 리스트 뷰가 구분될 수 있도록 변경했습니다.

Commit:
[fix] #79 Supabase에서 fetch 메소드에 쿼리 입력 추가

- RLS로 자신의 정보만 불러올 수 있도록 했지만, 메이트 정보도 불러와야 해서 테이블의 RLS 정책을 모든 사용자에게 열려있도록 바꿨습니다.
- 따라서 쿼리를 이용해서 정확한 메이트의 정보를 불러올 수 있또록 쿼리 입력 부분을 추가했습니다.

Commit:
[feat] #79 메이트 관련 유즈케이스 추가

- RequestMateInfoUseCase: 메이트의 정보를 서버에서 가져오는 역할을 합니다.
- RequestMateListUseCase: 유저의 메이트 목록을 서버에서 가져오는 역할을 합니다.

Commit:
[feat] #79 키워드 뷰 해시태그 로직 수정

키워드 뷰가 이니셜라이즈 될 때가 아닌 텍스트가 수정 될 때 #을 붙일 수 있도록 수정했습니다.

Commit:
[fix] #79 엔티티 이름 변경 적용

엔티티 이름이 변경된 것을 프로퍼티 이름에도 적용했습니다.

Commit:
[chore] #79 프로젝트 파일 업데이트

Commit:
[fix] #79 필요없는 메소드 호출 제거

pearhyunjin commented on SniffMeet/SniffMeet/Source/Walk/RequestWalk/View/ProfileView.swift:
```diff
@@ -69,14 +71,19 @@ final class ProfileView: BaseView {
+                keywordView.text = keyword
+                keywordView.isHidden = false
+            }
+//        guard let firstKeyword = keywords.first else { return }
+//        firstKeywordView.text = firstKeyword.rawValue
+//        
+//        guard mate.keywords.count > 1 else { return }
+//        secondKeywordView.text = mate.keywords[1].rawValue
```
p5. 지워도 되지 않을까요?

kelly-chui replied:
앗 까먹었네요 지우겠습니다!

pearhyunjin reviewed (approved):
굿입니당~ 수고하셨어요!

soletree commented on SniffMeet/SniffMeet/Source/Walk/RequestWalk/View/ProfileView.swift:
```diff
@@ -69,14 +71,19 @@ final class ProfileView: BaseView {
-        secondKeywordView.text = dog.keywords[1].rawValue
+    func configure(name: String, keywords: [String]) {
+        nameLabel.text = name
+
+        zip([firstKeywordView, secondKeywordView], keywords)
+            .forEach { keywordView, keyword in
```
zip하니까 깔끔하네요~~bb

soletree commented on SniffMeet/SniffMeet/Source/Walk/RequestWalk/Interactor/RequestWalkInteractor.swift:
```diff
@@ -5,25 +5,49 @@
 
+import Foundation
+
 protocol RequestWalkInteractable: AnyObject {
+    var mate: Mate { get }
     var presenter: RequestWalkInteractorOutput? { get set }
```
P5. mate는 presenter에서 명시적으로 접근하는 경우가 없어보여서 프로토콜로 추상화하지 않아도 괜찮을거 같은데, 혹시 이 부분 어떻게 생각하시나요??

kelly-chui replied:
바로 반영하겠습니다~~ 👍

soletree reviewed (approved):
고생하셨습니다~!! 👍👍

Commit:
[chore] 충돌 해결

Commit:
[fix] #79 필요 없는 코멘트, 추상화 제거

- 로직을 개선할때 기존 로직을 코멘트로 남겨놓은 부분을 제거했습니다.
- RequestWalkInteractable 프로토콜에서 mate를 가지고 있었지만, 외부에서 mate에 접근하지 않아 프로토콜에서 mate 프로퍼티를 제거했습니다.
